src/Jackalope/Version/VersionHandler.php

```diff
@@ -235,15 +236,203 @@ private function createFrozenNode(NodeInterface $versionNode, NodeInterface $nod
+        foreach ($frozenNode->getProperties() as $property) {
+            /** @var PropertyInterface $property */
+            $propertyName = $property->getName();
+            if ($propertyName == 'jcr:frozenPrimaryType'
+                || $propertyName == 'jcr:frozenMixinTypes'
+                || $propertyName == 'jcr:frozenUuid'
+            ) {
+                continue;
+            }
+
+            if ($node->hasProperty($propertyName)) {
+                $nodeProperty = $node->getProperty($propertyName);
+                $onParentVersion = $nodeProperty->getDefinition()->getOnParentVersion();
+                if ($onParentVersion == OnParentVersionAction::COPY || $onParentVersion == OnParentVersionAction::VERSION) {
+                    $nodeProperty->setValue($property->getValue());
+                }
+            } else {
+                // cannot check the onParentVersion attribute of a non existing property
+                // but if the property exists on the frozen node it has to be copy or version
+                $node->setProperty($propertyName, $property->getValue());
+            }
+        }
+
+        // handle properties present on the node but not on the frozen node
+        foreach ($node->getProperties() as $propertyName => $property) {
+            if ($frozenNode->hasProperty($propertyName)) {
+                continue;
+            }
+
+            $onParentVersion = $property->getDefinition()->getOnParentVersion();
+
+            switch ($onParentVersion) {
+                case OnParentVersionAction::COPY:
+                case OnParentVersionAction::VERSION:
+                case OnParentVersionAction::ABORT:
+                    $property->remove();
+                    break;
+                case OnParentVersionAction::INITIALIZE:
+                case OnParentVersionAction::COMPUTE: // implementation-specific, decided to handle as INITIALIZE
+                    // TODO how to get default value of property?
+            }
+        }
+
+        // TODO handle identifier collisions
+
+        // handle child nodes present on the frozen node
+        foreach ($frozenNode->getNodes() as $frozenChildNode) {
+            /** @var NodeInterface $frozenChildNode */
+            if (!$removeExisting) {
+                // TODO check occurence of node in repository
+                throw new NotImplementedException('Check for $removeExisting not implemented yet');
+            }
+
+            // TODO handle different onParentVersion values
+
+            $childNodePath = $node->getPath() . '/' . $frozenChildNode->getName();
+            if ($this->session->nodeExists($childNodePath)) {
+                $this->session->removeItem($childNodePath);
+            }
+
+            switch ($onParentVersion) {
+                case OnParentVersionAction::COPY:
+                    $this->restoreFromNode($node, $frozenChildNode);
+                    break;
+                case OnParentVersionAction::VERSION:
+
```

Comment:
if there is another version of the node currently valid in the workspace, the spec is clear. (depends on removeExisting). otherwise there seems to be a restoreByLabel, which would mean we try to find one with the same label in the history. i guess you meant 

"by a method call that does not specify a label then the workspace in which the restore is being performed will determine which particular version of C will be restored. This determination depends on the configuration of the workspace and is outside the scope of this specification."

i would suggest we just do the easiest thing in this case. can we find what version of the child was active at the point this version was created? if so, we should restore that, otherwise we should restore the last version i guess.